odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+
+
+class FunctionalComp(Functional, OperatorComp):
+    """Composition of a functional with an operator."""
+
+    def __init__(self, func, op, tmp1=None, tmp2=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The left ("outer") operator
+        op : `Operator`
+            The right ("inner") operator. Its range must coincide with the
+            domain of ``func``.
+        tmp1 : `element` of the range of ``op``, optional
+            Used to avoid the creation of a temporary when applying ``op``
+        tmp2 : `element` of the range of ``op``, optional
+            Used to avoid the creation of a temporary when applying the
+            gradient of ``func``
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is not a Functional instance.'
+                            ''.format(func))
+
+        OperatorComp.__init__(self, left=func, right=op, tmp=tmp1)
+
+        Functional.__init__(self, domain=func.domain,
+                            linear=(func.is_linear and op.is_linear),
+                            smooth=(func.is_smooth and op.is_linear),
+                            concave=(func.is_concave and op.is_linear),
+                            convex=(func.is_convex and op.is_linear),
+                            grad_lipschitz=np.infty)
+
+        if tmp2 is not None and tmp2 not in self._left.domain:
+            raise TypeError('second temporary {!r} not in the domain '
+                            '{!r} of the functional.'
+                            ''.format(tmp2, self._left.domain))
+        self._tmp2 = tmp2
+
+    @property
+    def gradient(self):
+        """Gradient of the compositon according to the chain rule."""
+
+        func = self.left
+        op = self.right
+
+        class CompositGradient(Operator):
+            """Gradient of the compositon according to the chain rule."""
+            def __init__(self):
+                """Creates an instance of CompositGradient."""
+                super().__init__(func.domain, func.domain, linear=False)
+                self._func = func
+                self._op = op
+
+            def _call(self, x):
+                """Applies the gradient operator to the given point.
+
+                Parameters
+                ----------
+                x : `LinearSpaceVector`
+                    Element in the domain of the functional to which the
+                    gradient operator is applied. The element must have a
+                    non-zero norm
+
+                Returns
+                -------
+                `self(x)` : `LinearSpaceVector`
+                    Evaluation of the gradient operator. An element in the
+                    domain of the functional.
+                """
+                return self._op.derivative(x).adjoint(
+                    self._func.gradient(self._op(x)))
+
+        return CompositGradient()
```

Comment:
In this case I don't think that works, and that we actually need an operator here which we can return.

If the operator that the composition is made with is not linear, I need a point to evaluate the derivative of the operator before I can take the adjoint of it. But `gradient` is a property, so I don't know that point when I return the operator.